cats/cats-sql/src/main/kotlin/com/netflix/spinnaker/cats/sql/cache/SqlCache.kt

```diff
@@ -884,15 +885,51 @@ class SqlCache(
   }
 
   private fun resourceTableName(type: String): String =
-    "cats_v${schemaVersion}_${if (tableNamespace != null) "${tableNamespace}_" else ""}${sanitizeType(type)}"
+    checkTableName("cats_v${schemaVersion}_", sanitizeType(type), "")
 
   private fun relTableName(type: String): String =
-    "cats_v${schemaVersion}_${if (tableNamespace != null) "${tableNamespace}_" else ""}${sanitizeType(type)}_rel"
+    checkTableName("cats_v${schemaVersion}_", sanitizeType(type), "_rel")
 
   private fun sanitizeType(type: String): String {
     return type.replace(typeSanitization, "_")
   }
 
+  /**
+   * Computes the actual name of the table less than MAX_TABLE_NAME_LENGTH characters long.
+   * It always keeps prefix with tablenamespace but can shorten name and suffix in that order.
+   * @return computed table name
+   */
+  private fun checkTableName(prefix: String, name: String, suffix: String): String {
+    var base = prefix
+    if (tableNamespace != null) {
+      base = "${prefix + tableNamespace}_"
+    }
+
+    // Optimistic and most frequent case
+    val tableName = base + name + suffix
+    if (tableName.length < sqlConstraints.maxTableNameLength) {
+      return tableName
+    }
+
+    // Hash the name and keep the suffix
+    val hash = Hashing.murmur3_128().hashBytes(name.toByteArray()).toString().substring(0..15)
+    val available = sqlConstraints.maxTableNameLength - base.length - suffix.length - hash.length - 1
+    if (available >= 0) {
+      return base + name.substring(0..available) + hash + suffix
+    }
+
+    // Hash name and suffix
+    val pHash = Hashing.murmur3_128().hashBytes((name + suffix).toByteArray()).toString().substring(0..15)
```

Comment:
Avoid hashing twice in this case. They're all known up front so determine what to hash based on the length of the relevant components.

Reply:
Thanks, Updated